File: apps/apple/core/Sources/LorvexDomain/Timezone.swift
import Foundation

/// Timezone parsing / normalization helpers and the timezone-aware
/// "today" / "today + N days" `YYYY-MM-DD` helpers used for day-boundary logic.
///
/// Generic event/reminder fields accept any identifier Foundation resolves.
/// The synced product-timezone preference is stricter: it uses Apple's canonical
/// region list (plus canonical `UTC`) so aliases and fixed-offset spellings do
/// not create distinct synced bytes for the same calendar-day authority.
/// A failure resolving the anchored timezone, carrying the human-readable
/// message used by the anchored-timezone resolver.
public struct TimezoneResolutionError: Error, Equatable, CustomStringConvertible {
  public let message: String
  public init(_ message: String) { self.message = message }
  public var description: String { message }
}

public enum Timezone {
  /// Parse any non-empty timezone identifier Foundation recognizes.
  public static func parseTimezoneName(_ value: String) -> TimeZone? {
    let trimmed = ValidationFormat.trimWhitespace(value)
    guard !trimmed.isEmpty else { return nil }
    return TimeZone(identifier: trimmed)
  }

  /// Normalize a generic Foundation timezone identifier. Event and reminder
  /// payloads use this broader contract because fixed offsets are valid there.
  public static func normalizeTimezoneName(_ value: String?) -> String? {
    guard let value = value else { return nil }
    let trimmed = ValidationFormat.trimWhitespace(value)
    return parseTimezoneName(trimmed) != nil ? trimmed : nil
  }

  /// Normalize the synced product-day authority to one stable identifier.
  /// Abbreviations (`PST`), fixed offsets (`GMT+5`), and alias spellings absent
  /// from Apple's canonical identifier list are rejected. `GMT` and `UTC` both
  /// normalize to the single product spelling `UTC`.
  public static func normalizeProductTimezoneName(_ value: String?) -> String? {
    guard let value else { return nil }
    let trimmed = ValidationFormat.trimWhitespace(value)
    guard !trimmed.isEmpty else { return nil }
    if trimmed == "UTC" || trimmed == "GMT" { return "UTC" }
    guard TimeZone.knownTimeZoneIdentifiers.contains(trimmed) else { return nil }
    return trimmed
  }

  /// Parse a canonical JSON string preference containing a canonical product timezone.
  /// Raw unquoted strings and invalid identifiers are rejected.
  public static func parseJsonTimezonePreference(_ raw: String?) -> String? {
    guard let timezone = Parsing.parseJsonStringPreference(raw) else { return nil }
    return normalizeProductTimezoneName(timezone)
  }

  /// Parse a stored timezone preference that must contain a valid canonical IANA
  /// timezone string, surfacing a typed ``ValidationError`` for the missing /
  /// invalid cases.
  public static func parseRequiredTimezonePreference(
    _ raw: String, key: String
  ) -> Result<String, ValidationError> {
    guard let timezone = Parsing.parseJsonStringPreference(raw) else {
      return .failure(
        .message("invalid \(key) preference: expected canonical JSON timezone string"))
    }
    if let canonical = normalizeProductTimezoneName(timezone) {
      return .success(canonical)
    }
    return .failure(.message("invalid \(key) preference: unknown timezone '\(timezone)'"))
  }

  /// Resolve the anchored timezone used for calendar/day-boundary logic.
  ///
  /// Preference order: an explicit validated active timezone, then the current
  /// system IANA timezone (when resolvable and valid). When neither is available
  /// the caller fails rather than silently substituting UTC, which would shift
  /// calendar-day boundaries. The system lookup is passed in as a `Result` so
  /// the domain layer stays free of implicit host-clock reads.
  public static func resolveAnchoredTimezoneName(
    activeTimezone: String?,
    systemTimezoneLookup: Result<String, TimezoneResolutionError>
  ) -> Result<String, TimezoneResolutionError> {
    if let activeTimezone,
      let canonical = normalizeProductTimezoneName(activeTimezone)
    {
      return .success(canonical)
    }
    switch systemTimezoneLookup {
    case let .failure(error):
      return .failure(
        TimezoneResolutionError(
          "anchored timezone requires a resolvable system IANA timezone: \(error.message)"))
    case let .success(timezone):
      if let normalized = normalizeProductTimezoneName(timezone) {
        return .success(normalized)
      }
      return .failure(
        TimezoneResolutionError(
          "anchored timezone requires a valid IANA timezone, got '\(timezone)'"))
    }
  }

  /// Today's date as `YYYY-MM-DD` in the given IANA timezone, falling back to
  /// `systemFallback` when the timezone name is `nil` or unrecognized.
  ///
  /// `now` is the reference instant (the UTC wall clock). `systemFallback` is the
  /// time zone to use when no valid preference is supplied — the caller passes its
  /// system-local zone, keeping the domain layer free of implicit host reads.
  /// An invalid (`Some(invalid)`) preference name falls back gracefully to
  /// `systemFallback` rather than aborting.
  public static func todayYmdForTimezoneName(
    now: Date, timezoneName: String?, systemFallback: TimeZone
  ) -> String {
    baseDateComponents(now: now, timezoneName: timezoneName, systemFallback: systemFallback)
      .canonicalString
  }

  /// Today + `offsetDays` as `YYYY-MM-DD` in the given IANA timezone, falling back
  /// to `systemFallback` when the name is `nil` or unrecognized.
  public static func datePlusDaysYmdForTimezoneName(
    now: Date, timezoneName: String?, offsetDays: Int, systemFallback: TimeZone
  ) -> String {
    let base = baseDateComponents(now: now, timezoneName: timezoneName, systemFallback: systemFallback)
    // Day arithmetic on a pure date is timezone-agnostic; the shared UTC
    // calendar means adding days never crosses a DST boundary in the math.
    let cal = IsoDate.calendar
    var dc = DateComponents()
    dc.year = base.year
    dc.month = base.month
    dc.day = base.day
    guard let date = cal.date(from: dc),
      let shifted = cal.date(byAdding: .day, value: offsetDays, to: date)
    else { return base.canonicalString }
    let c = cal.dateComponents([.year, .month, .day], from: shifted)
    return IsoDate.YMD(year: c.year ?? 0, month: c.month ?? 0, day: c.day ?? 0).canonicalString
  }

  /// The first midnight strictly after `now` in the configured product zone.
  ///
  /// Long-lived app processes use this boundary to refresh their materialized
  /// logical day even when the device itself is in another time zone (where
  /// `NSCalendarDayChanged` fires at a different instant). Invalid names fail
  /// closed instead of silently scheduling against the device zone.
  public static func nextMidnight(after now: Date, timezoneName: String) -> Date? {
    guard let zone = parseTimezoneName(timezoneName) else { return nil }
    var calendar = Calendar(identifier: .gregorian)
    calendar.locale = Locale(identifier: "en_US_POSIX")
    calendar.timeZone = zone
    return calendar.nextDate(
      after: now,
      matching: DateComponents(hour: 0, minute: 0, second: 0),
      matchingPolicy: .nextTime,
      direction: .forward
    )
  }

  /// Resolve the calendar date `now` falls on under `timezoneName`, falling back
  /// to `systemFallback` on either `nil` or an unparseable name.
  static func baseDateComponents(
    now: Date, timezoneName: String?, systemFallback: TimeZone
  ) -> IsoDate.YMD {
    let zone: TimeZone
    if let name = timezoneName, let parsed = parseTimezoneName(name) {
      zone = parsed
    } else {
      zone = systemFallback
    }
    var cal = Calendar(identifier: .gregorian)
    cal.locale = Locale(identifier: "en_US_POSIX")
    cal.timeZone = zone
    let c = cal.dateComponents([.year, .month, .day], from: now)
    return IsoDate.YMD(year: c.year ?? 0, month: c.month ?? 0, day: c.day ?? 0)
  }
}

File: apps/tauri/lorvex-domain/src/time/timezone.rs
//! IANA timezone parsing / normalization helpers and the timezone-aware
//! "today" / "today + N days" `YYYY-MM-DD` helpers used for
//! day-boundary logic across the workspace.

use chrono::{DateTime, Duration, Local, NaiveDate, Utc};
use chrono_tz::Tz;

/// Parse and validate a non-empty IANA timezone name.
pub fn parse_timezone_name(value: &str) -> Option<Tz> {
    let trimmed = value.trim();
    (!trimmed.is_empty())
        .then_some(trimmed)
        .and_then(|timezone| timezone.parse::<Tz>().ok())
}

/// Normalize a timezone name only if it is a valid IANA timezone.
///
/// Returns the trimmed string when valid; otherwise `None`.
pub fn normalize_timezone_name(value: Option<&str>) -> Option<String> {
    let trimmed = value?.trim();
    parse_timezone_name(trimmed).map(|_| trimmed.to_string())
}

/// Parse a canonical JSON string preference containing a valid IANA timezone.
///
/// Raw unquoted fallback strings and invalid timezone identifiers are rejected.
pub fn parse_json_timezone_preference(raw: Option<&str>) -> Option<String> {
    let timezone = crate::parsing::parse_json_string_preference(raw)?;
    parse_timezone_name(&timezone)?;
    Some(timezone)
}

/// Parse a stored timezone preference that must contain a valid canonical IANA
/// timezone string. Returns a typed `ValidationError` so consumers can
/// `?`-propagate through the existing `From<ValidationError>` impls
/// instead of stringifying at every boundary (#3288).
pub fn parse_required_timezone_preference(
    raw: &str,
    key: &str,
) -> Result<String, crate::validation::ValidationError> {
    let timezone = crate::parsing::parse_json_string_preference(Some(raw)).ok_or_else(|| {
        crate::validation::ValidationError::Message(format!(
            "invalid {key} preference: expected canonical JSON timezone string"
        ))
    })?;
    if parse_timezone_name(&timezone).is_some() {
        Ok(timezone)
    } else {
        Err(crate::validation::ValidationError::Message(format!(
            "invalid {key} preference: unknown timezone '{timezone}'"
        )))
    }
}

/// Resolve the anchored timezone used for calendar/day-boundary logic.
///
/// Preference order:
/// 1. Explicit validated active timezone from preferences.
/// 2. Current system IANA timezone, if resolvable and valid.
///
/// When neither source is available, callers should fail instead of silently
/// substituting `UTC`, because that would shift calendar-day boundaries and
/// mis-anchor local-time queries.
pub fn resolve_anchored_timezone_name(
    active_timezone: Option<String>,
    system_timezone_lookup: Result<String, String>,
) -> Result<String, String> {
    if let Some(active_timezone) = active_timezone {
        return Ok(active_timezone);
    }

    let timezone = system_timezone_lookup.map_err(|error| {
        format!("anchored timezone requires a resolvable system IANA timezone: {error}")
    })?;
    normalize_timezone_name(Some(&timezone)).ok_or_else(|| {
        format!("anchored timezone requires a valid IANA timezone, got '{timezone}'")
    })
}

// the bare `local_today_ymd` /
// `local_date_plus_days_ymd` helpers live here. They
// silently routed through `chrono::Local` (the host machine's
// timezone) regardless of the user's stored `PREF_TIMEZONE`, so a
// user with `America/Los_Angeles` running on a UTC-set CI/server
// box saw "today" against the host clock instead of their own —
// off by up to a calendar day for every "today / tomorrow /
// yesterday" lookup. Deleting the helpers forces every caller to
// pass an explicit timezone name (which falls back to system-local
// only when the caller passes `None`, i.e. legitimately has no
// preference). Callers that need a fallback chain (active
// preference, then system) should route through
// [`resolve_anchored_timezone_name`] so the failure surfaces as a
// typed error instead of a silent zone substitution.

/// Today's date as `YYYY-MM-DD` in the given IANA timezone, falling back to
/// system-local when the timezone name is `None` or unrecognized.
///
/// a *corrupt* (Some(invalid)) timezone preference is
/// distinct from an *unset* (None) one — `None` legitimately means
/// "no preference, use system-local", but `Some("America/Not_A_Zone")`
/// is a data-integrity failure (preference table corruption,
/// truncated import, manual DB edit). In dev builds the
/// `debug_assert!` panics so the test suite catches the case at the
/// failure site; in release we still fall back gracefully because
/// the alternative (panicking on every today-query) is worse than
/// returning a possibly-off-by-one date.
pub fn today_ymd_for_timezone_name(now: DateTime<Utc>, timezone_name: Option<&str>) -> String {
    base_date_in_timezone(now, timezone_name, "today_ymd_for_timezone_name")
        .format("%Y-%m-%d")
        .to_string()
}

/// Today + `offset_days` as `YYYY-MM-DD` in the given IANA timezone, falling
/// back to system-local when the timezone name is `None` or unrecognized.
/// See [`today_ymd_for_timezone_name`] for the corrupt-preference contract.
pub fn date_plus_days_ymd_for_timezone_name(
    now: DateTime<Utc>,
    timezone_name: Option<&str>,
    offset_days: i64,
) -> String {
    let base_date =
        base_date_in_timezone(now, timezone_name, "date_plus_days_ymd_for_timezone_name");
    (base_date + Duration::days(offset_days))
        .format("%Y-%m-%d")
        .to_string()
}

/// Resolve the calendar date `now` falls on under `timezone_name`,
/// falling back to system-local on either `None` or an unparseable
/// timezone string. Mirrors the corrupt-preference contract documented
/// on [`today_ymd_for_timezone_name`]: an explicit non-empty but
/// invalid IANA name fires a `debug_assert!` so test runs catch the
/// drift while release builds still degrade gracefully.
///
/// `caller` names the caller for the assertion message; in release it
/// is unused and compiles away with the `debug_assert!`.
fn base_date_in_timezone(
    now: DateTime<Utc>,
    timezone_name: Option<&str>,
    caller: &'static str,
) -> NaiveDate {
    if let Some(name) = timezone_name {
        if let Some(tz) = parse_timezone_name(name) {
            return now.with_timezone(&tz).date_naive();
        }
        debug_assert!(
            name.is_empty(),
            "{caller}: invalid IANA timezone '{name}'; falling back to system-local"
        );
    }
    now.with_timezone(&Local).date_naive()
}
